File: src/main/java/refactor/algorithm/Refactor.java
package refactor.algorithm;

import org.apache.commons.collections4.ListUtils;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Refactor {
    private final Map<String, Cell> livingCells;

    Refactor() {
        livingCells = Collections.emptyMap();
    }

    public Refactor(Map<String, Cell> livingCells) {
        this.livingCells = livingCells;
    }

    Cell findCellWithLeastWeight() {
        return getLivingCells().stream().min(Comparator.comparingInt(Cell::getWeight)).orElse(null);
    }

    public boolean isLivingCell(int x, int y) {
        return isLivingCell(Cell.toName(x, y));
    }

    private boolean isLivingCell(String key) {
        return livingCells.get(key) != null;
    }

    Collection<Cell> getLivingCells() {
        if (livingCells.isEmpty()) {
            throw new RuntimeException("No more living cells");
        }
        return livingCells.values();
    }

    public Refactor tick() {
        return new Refactor(getNextMap());
    }

    @NotNull
    private Map<String, Cell> getNextMap() {
        return Stream.concat(getNextGenerationCells(), getReproducibleCells())
                .collect(Collectors.toMap(Cell::toString, cell -> cell));
    }

    private Stream<Cell> getNextGenerationCells() {
        return getFilteredCells(getLivingCells().stream(), n -> n == 2 || n == 3);
    }

    private Stream<Cell> getReproducibleCells() {
        return getFilteredCells(getNeighbouringDeadCells(), n -> n == 3);
    }

    private Stream<Cell> getFilteredCells(Stream<Cell> list, Predicate<Long> filter) {
        return list.filter(cell -> filter.test(getNumberOfLivingNeighbours(cell)));
    }

    private long getNumberOfLivingNeighbours(Cell that) {
        return that.getNeighbours().stream().filter(cell -> isLivingCell(cell.toString())).count();
    }

    Stream<Cell> getNeighbouringDeadCells() {
        return getLivingCells().stream().flatMap(Cell::getNeighbors).filter(this::isDeadCell).distinct();
    }

    private boolean isDeadCell(Cell cell) {
        return !isLivingCell(cell.toString());
    }
}

File: src/main/kotlin/refactor/algorithm/Refactor.kt
package refactor.algorithm

import refactor.algorithm.Cell.Companion.toName

class Refactor(private val livingCellsMap: Map<String, Cell>) {
    fun isLivingCell(x: Int, y: Int) = toName(x, y).isLivingCell
    val livingCells get() = livingCellsMap.values
    fun tick() = Refactor(nextMap)
    val neighbouringDeadCells get() = livingCells.flatMap{ it.neighbours }.filter{ it.isDeadCell }.distinct()

    private val String.isLivingCell get() = livingCellsMap[this] is Cell
    private val nextMap get() = (nextGenerationCells + reproducibleCells).map{(it.name to it)}.toMap()
    private val nextGenerationCells get() = livingCells.getFilteredCells{ it in 2..3 }
    private val reproducibleCells get() = neighbouringDeadCells.getFilteredCells{ it == 3 }
    private fun Collection<Cell>.getFilteredCells(test: (Int) -> Boolean) = filter { test(it.numberOfLivingNeighbours) }
    private val Cell.numberOfLivingNeighbours get() = neighbours.filter { it.name.isLivingCell }.count()
    private val Cell.isDeadCell get() = !name.isLivingCell
}